V0.12.0.x use specific proto and fix sync issues

UdjinM6 opened the pull request:
- store all MNs that fit min payment proto, use only those who fit specific proto (pool, ix, budget)
- ~~use ```pindexBestHeader``` height to identify recent block instead of _magic_ time~~ <--- doesn't play nice with forks
- reset ```last...``` on sync initial state
- update ```lastMasternodeList``` when we received legit update for MN we already know
- ask for ```mnb``` on ```mnp``` even if not synced yet
- change rpc and "tools -> info" to show "_ds / enabled / all_" for ```count```

UdjinM6 commented:
Don't merge it yet, having issues syncing during forks

evan82 commented on src/masternode-sync.cpp:
```diff
@@ -74,43 +83,36 @@ void CMasternodeSync::Process()
-            //don't begin syncing until we're at a recent block
-            if(pindexPrev->nTime + 600 < GetTime()) return;
+        //don't begin syncing until we're almost at a recent block
+        if(pindexPrev->nHeight + 4 < pindexBestHeader->nHeight || pindexPrev->nTime + 600 < GetTime()) return;
```
Shouldn't we wait to ask for mnb/mbp until we have the most recent block? Otherwise if a masternodes input is newer than the current block, we will reject it and have inconsistancies

UdjinM6 replied:
Well, I agree. I'm not sure on that either tbh. I just divided 600 by blocktime (i.e. 150) and that's where 4 came from :)
On another note we are just entering cycle to ask other nodes for MNs so most likely we'll be already at a recent block when the first mnb from them arrive.

evan82 replied:
Gotcha, shouldn't it be "pindexPrev->nHeight + 4 > pindexBestHeader->nHeight" though? Being that we get the headers first and we're checking to see when the chain actually downloads, right?

UdjinM6 replied:
err... nope. Say we have 100 headers and only 10 blocks so
```if (10 + 4 < 100 || ...) return; // do not sync yet```
Or am I missing smth?
I would agree however that it's better to use ```&&``` here
```if(pindexPrev->nHeight + 4 < pindexBestHeader->nHeight && pindexPrev->nTime + 600 < GetTime()) return;```
i.e. if block height is still low AND block time is low too THEN do nothing. Thoughts?

darkcoinproject replied:
Oh, I just was reading it backwards somehow :) I think using && is probably
a lot better in that case

On Sun, Jul 19, 2015 at 7:46 AM, UdjinM6 <notifications@github.com> wrote:

> In src/masternode-sync.cpp
> <https://github.com/dashpay/dash/pull/441#discussion_r34958712>:
>
> >
> > -            //don't begin syncing until we're at a recent block
> > -            if(pindexPrev->nTime + 600 < GetTime()) return;
> > +        //don't begin syncing until we're almost at a recent block
> > +        if(pindexPrev->nHeight + 4 < pindexBestHeader->nHeight || pindexPrev->nTime + 600 < GetTime()) return;
>
> err... nope. Say we have 100 headers and only 10 blocks so
> if (10 + 4 < 100 || ...) return; // do not sync yet
> Or am I missing smth?
> I would agree however that it's better to use && here
> if(pindexPrev->nHeight + 4 < pindexBestHeader->nHeight &&
> pindexPrev->nTime + 600 < GetTime()) return;
> i.e. if block height is still low AND block time is low too THEN do
> nothing. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dashpay/dash/pull/441/files#r34958712>.
>

UdjinM6 replied:
yeah, np, happens to me too sometimes :smile: 
changed to ```&&```, see next commit

UdjinM6 commented:
rebased

UdjinM6 commented:
Ok, rebased again and also removed ```-masternodeminprotocol``` and ```nMasternodeMinProtocol```
It's better to use ```masternodePayments.GetMinMasternodePaymentsProto()``` (it is used in ```CountEnabled()``` by default now) to ensure everyone is on the same page.
